File: src/libmojave-external/cutil/lm_ssl.c
/*
 * Interface to OpenSSL.  This looks just like a standard socket
 * API, except that a few of the functions require certificates.
 *
 * ----------------------------------------------------------------
 *
 * @begin[license]
 * Copyright (C) 2003 Mojave Group, Caltech
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation,
 * version 2.1 of the License.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 * 
 * Additional permission is given to link this library with the
 * OpenSSL project's "OpenSSL" library, and with the OCaml runtime,
 * and you may distribute the linked executables.  See the file
 * LICENSE.libmojave for more details.
 *
 * Author: Jason Hickey
 * @email{jyh@cs.caltech.edu}
 * @end[license]
 */
#include <stdio.h>
#include <memory.h>
#include <stdlib.h>
#include <caml/mlvalues.h>
#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/fail.h>
#include <caml/custom.h>

#ifdef WIN32
#  include <winsock.h>
#  define SHUT_RD 0
#  define SHUT_WR 1

typedef long socklen_t;
#else /* !WIN32 */
#  include <unistd.h>
#  include <sys/types.h>
#  include <sys/socket.h>
#  include <netinet/in.h>
#  include <fcntl.h>

typedef int SOCKET;

#define closesocket close
#endif /* !WIN32 */

#define Nothing ((value) 0)

extern void uerror (char * cmdname, value arg) Noreturn;

#define uerr(cmd) uerror(cmd, Nothing)

void enter_blocking_section(void);
void leave_blocking_section(void);

/*
 * Inet addresses are strings.
 */
static value alloc_inet_addr(uint32 addr)
{
    value a;
    a = alloc_string(sizeof(uint32));
    *(uint32 *)a = addr;
    return a;
}

#ifdef SSL_ENABLED

/************************************************************************
 * SSL routines.
 */

#include <openssl/ssl.h>
#include <openssl/err.h>
#include <openssl/evp.h>

/*
 * Struct for the socket.
 *   server: is this a server socket?
 *   fd: file descriptor for the socket
 *   context: the SSL context
 *   ssl:
 */
typedef enum {
    SSL_INFO_MASTER,
    SSL_INFO_SLAVE
} SslInfoKind;

typedef struct {
    SslInfoKind kind;
    SOCKET fd;
    SSL_CTX *context;
    SSL *ssl;
} SslInfo;

/************************************************
 * ML interface to addresses.
 */

/*
 * Print an error message.
 */
static void print_errors()
{
    char buffer[1024];
    unsigned long err;
    const char *s;

    while((err = ERR_get_error()))
        fprintf(stderr, "lm_ssl: %s\n", ERR_error_string(err, buffer));
}

/*
 * Create a new context.
 */
static SSL_CTX *lm_ssl_ctx_new(const char *keyfile)
{
    SSL_CTX *context;

    /* Create a new context and initialize */
    context = SSL_CTX_new(SSLv23_method());
    if(context == 0) {
        print_errors();
        failwith("lm_ssl_ctx_new: can't allocate context");
    }

    /* Load our certificate */
    if(SSL_CTX_use_certificate_chain_file(context, keyfile) == 0) {
        fprintf(stderr, "keyfile: %s\n", keyfile);
        print_errors();
        failwith("lm_ssl_ctx_new: can't assign certificate");
    }

    /* Password */
    if(SSL_CTX_use_PrivateKey_file(context, keyfile, SSL_FILETYPE_PEM) == 0) {
        fprintf(stderr, "keyfile: %s\n", keyfile);
        print_errors();
        failwith("lm_ssl_ctx_new: can't set private key");
    }

    return context;
}

/*
 * Add the DH file to the context.
 */
static void lm_ssl_ctx_dhfile(SSL_CTX *context, const char *dhfile)
{
    BIO *bio;
    DH *ret;

    /* Set the DH parameters */
    bio = BIO_new_file(dhfile, "r");
    if(bio == 0) {
        print_errors();
        failwith("lm_ssl_ctx_new: can't open DH file");
    }

    /* Read the file */
    ret = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
    BIO_free(bio);
    if(SSL_CTX_set_tmp_dh(context, ret) < 0) {
        print_errors();
        failwith("lm_ssl_ctx_new: can't set DH params");
    }
}

/*
 * Custom blocks.
 */
#define SslInfo_val(v)   ((SslInfo *) Data_custom_val(v))

static int ssl_info_compare(value v1, value v2)
{
    SslInfo *info1 = SslInfo_val(v1);
    SslInfo *info2 = SslInfo_val(v2);

    return info1->fd == info2->fd ? 0 : info1->fd < info2->fd ? -1 : 1;
}

static long ssl_info_hash(value v)
{
    return (long) SslInfo_val(v);
}

static void ssl_finalize(value v_info)
{
    SslInfo *info;
    int code;

    info = SslInfo_val(v_info);
    if(info->ssl) {
        SSL_free(info->ssl);
        info->ssl = 0;
    }
    if(info->kind == SSL_INFO_MASTER && info->context) {
        SSL_CTX_free(info->context);
        info->context = 0;
    }
    if(info->fd >= 0) {
        closesocket(info->fd);
        info->fd = -1;
    }
}

/*
 * Pass info in a custom block.
 */
static struct custom_operations win_handle_ops = {
    "ssl_socket",
    ssl_finalize,
    ssl_info_compare,
    ssl_info_hash,
    custom_serialize_default,
    custom_deserialize_default
};

static value ssl_info_new(SslInfoKind kind, SOCKET fd, SSL_CTX *context, SSL *ssl)
{
    value v = alloc_custom(&win_handle_ops, sizeof(SslInfo), 0, 1);
    SslInfo *info = SslInfo_val(v);
    info->kind = kind;
    info->fd = fd;
    info->context = context;
    info->ssl = ssl;
    return v;
}

/************************************************************************
 * Public functions.
 */

/*
 * Say whether SSL is enabled.
 */
value lm_ssl_enabled(value x)
{
    return Val_true;
}

/*
 * Start the SSL server.
 */
value lm_ssl_init(value x)
{
    SSL_load_error_strings();
    SSL_library_init();
    return Val_unit;
}

/*
 * Create a new socket.
 */
value lm_ssl_socket(value v_keyfile)
{
    SOCKET s;
    SSL_CTX *context;
    int one = 1;

    /* Add context */
    context = lm_ssl_ctx_new(String_val(v_keyfile));

    /* Open the socket */
    s = socket(PF_INET, SOCK_STREAM, 0);
    if(s < 0) {
        SSL_CTX_free(context);
        failwith("lm_ssl_socket: socket call failed");
    }

    /* Allow the address to be reused */
    setsockopt(s, SOL_SOCKET, SO_REUSEADDR, (void *) &one, sizeof(one));

    return ssl_info_new(SSL_INFO_MASTER, s, context, 0);
}

/*
 * Reuse an existing socket.
 */
value lm_ssl_serve(value v_fd, value v_keyfile, value v_dhfile)
{
    SOCKET s;
    SSL_CTX *context;
    SslInfo *info;
    value v_info;
    int one = 1;

    /* Add context */
    context = lm_ssl_ctx_new(String_val(v_keyfile));

    /* Open the socket */
    s = Int_val(v_fd);

    /* Allow the address to be reused */
    setsockopt(s, SOL_SOCKET, SO_REUSEADDR, (void *) &one, sizeof(one));

    /* Allocate the context */
    v_info = ssl_info_new(SSL_INFO_MASTER, s, context, 0);
    info = SslInfo_val(v_info);

    /* Set the Diffie-Helman file */
    lm_ssl_ctx_dhfile(info->context, String_val(v_dhfile));

    return v_info;
}

/*
 * Get an integer describing the socket.
 */
value lm_ssl_fd(value v_info)
{
    SslInfo *info = SslInfo_val(v_info);
    return Val_int(info->fd);
}

/*
 * Bind the socket to an address.
 */
value lm_ssl_bind(value v_info, value v_addr, value v_port)
{
    SslInfo *info;
    struct sockaddr_in sin;

    /* Get the address */
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = *(uint32 *)v_addr;
    sin.sin_port = htons((short) Int_val(v_port));

    /* Perform the bind */
    info = SslInfo_val(v_info);
    if(bind(info->fd, (struct sockaddr *) &sin, sizeof(sin)) < 0) {
		  uerr("lm_ssl_bind");
    }

    return Val_unit;
}

/*
 * Get the address of the socket.
 */
value lm_ssl_get_addr(value v_info)
{
    SslInfo *info;
    struct sockaddr_in sin;
    socklen_t size;
    value a, addr;

    /* Get the address */
    info = SslInfo_val(v_info);
    size = sizeof(sin);
    if(getsockname(info->fd, (struct sockaddr *) &sin, &size) < 0 || sin.sin_family != AF_INET) {
        uerr("lm_ssl_get_addr: getsockname");
    }

    /* Allocate the address */
    a = alloc_inet_addr(sin.sin_addr.s_addr);
    Begin_root(a);
    addr = alloc_small(2, 0);
    Field(addr, 0) = a;
    Field(addr, 1) = Val_int(ntohs(sin.sin_port));
    End_roots();
    return addr;
}

/*
 * Listen.
 */
value lm_ssl_listen(value v_info, value v_dhfile, value v_count)
{
    SslInfo *info;

    info = SslInfo_val(v_info);
    lm_ssl_ctx_dhfile(info->context, String_val(v_dhfile));
    listen(info->fd, Int_val(v_count));
    return Val_unit;
}

/*
 * Accept a connection on the socket.
 */
value lm_ssl_accept(value v_info)
{
    SslInfo *info;
    int fd, code;
    SSL *ssl;

    /* Listen on the server socket */
    info = SslInfo_val(v_info);
    enter_blocking_section();
    fd = accept(info->fd, (struct sockaddr *) 0, 0);
    leave_blocking_section();
    if(fd < 0) {
        uerr("lm_ssl_accept");
    }

#ifndef WIN32
    /* Set the close-on-exec flag */
    fcntl(fd, F_SETFD, FD_CLOEXEC);
#endif

    /* Start SSL operations */
    ssl = SSL_new(info->context);
    if(ssl == 0) {
        closesocket(fd);
        print_errors();
        failwith("lm_ssl_accept");
    }

    /* Set the descriptor */
    code = SSL_set_fd(ssl, fd);
    if(code <= 0) {
        closesocket(fd);
        SSL_free(ssl);
        print_errors();
        failwith("lm_ssl_accept: set_fd failed");
    }

    /* Negotiate */
    code = SSL_accept(ssl);
    if(code <= 0) {
        closesocket(fd);
        SSL_free(ssl);
        print_errors();
        failwith("lm_ssl_accept: negotiation failed");
    }

    /* Allocate a new struct */
    return ssl_info_new(SSL_INFO_SLAVE, fd, info->context, ssl);
}

/*
 * Make a connection on the socket.
 */
value lm_ssl_connect(value v_info, value v_addr, value v_port)
{
    SslInfo *info;
    struct sockaddr_in sin;
    int fd, code;
    SSL *ssl;

    /* Check to make sure not connected */
    info = SslInfo_val(v_info);
    if(info->ssl)
        failwith("lm_ssl_connect: already connected");

    /* Get the address */
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = *(uint32 *)v_addr;
    sin.sin_port = htons((short) Int_val(v_port));

    /* Make the connection */
    enter_blocking_section();
    code = connect(info->fd, (struct sockaddr *) &sin, sizeof(sin));
    leave_blocking_section();
    if(code < 0) {
        uerr("lm_ssl_connect");
    }

    /* Start SSL operations */
    ssl = SSL_new(info->context);
    if(ssl == 0) {
        print_errors();
        failwith("lm_ssl_connect");
    }

    /* Set the descriptor */
    code = SSL_set_fd(ssl, info->fd);
    if(code <= 0) {
        SSL_free(ssl);
        print_errors();
        failwith("lm_ssl_connect: set_fd failed");
    }

    /* Negotiate */
    code = SSL_connect(ssl);
    if(code <= 0) {
        SSL_free(ssl);
        print_errors();
        failwith("lm_ssl_connect: negotiation failed");
    }

    info->ssl = ssl;
    return Val_unit;
}

/*
 * Read some data from the connection.
 */
value lm_ssl_read(value v_info, value v_string, value v_off, value v_len)
{
    int off, len, amount;
    SslInfo *info;
    char *buf;

    info = SslInfo_val(v_info);
    buf = String_val(v_string);
    off = Int_val(v_off);
    len = Int_val(v_len);
    enter_blocking_section();
    amount = SSL_read(info->ssl, buf + off, len);
    leave_blocking_section();
    return Val_int(amount);
}

/*
 * Write some data to the connection.
 */
value lm_ssl_write(value v_info, value v_string, value v_off, value v_len)
{
    int off, len, amount;
    SslInfo *info;
    char *buf;

    info = SslInfo_val(v_info);
    buf = String_val(v_string);
    off = Int_val(v_off);
    len = Int_val(v_len);
    enter_blocking_section();
    amount = SSL_write(info->ssl, buf + off, len);
    leave_blocking_section();
    return Val_int(amount);
}

/*
 * Flush the connection.
 */
value lm_ssl_flush(value v_info)
{
    SslInfo *info;
    BIO *bio;

    info = SslInfo_val(v_info);
    enter_blocking_section();
    bio = SSL_get_wbio(info->ssl);
    if(bio)
        /* XXX: BUG! 
         * According to the man page:
         * BIO_flush(), because it can write data may return 0 or -1 indicating that the call should be retried later
         * in a similar manner to BIO_write().  The BIO_should_retry() call should be used and appropriate action
         * taken is the call fails.
         */
        BIO_flush(bio);
    leave_blocking_section();
    return Val_unit;
}

/*
 * Shutdown the connection.
 */
value lm_ssl_shutdown(value v_info)
{
    SslInfo *info;
    int code;

    /* Shutdown the connection */
    info = SslInfo_val(v_info);
    code = SSL_shutdown(info->ssl);

    /*
     * If we called shutdown first, the return code is always 0.
     * In this case, shutdown the TCP connection
     * and try again.
     */
    if(code == 0) {
        shutdown(info->fd, SHUT_WR);
        code = SSL_shutdown(info->ssl);
    }

    /* SSL struct is no longer needed */
    SSL_free(info->ssl);
    info->ssl = 0;
    return Val_unit;
}

/*
 * Close the connection.
 */
value lm_ssl_close(value v_info)
{
    ssl_finalize(v_info);
    return Val_unit;
}

#else /* !SSL_ENABLED */

/*
 * Unencrypted socket connection.
 */
typedef struct {
    SOCKET fd;
} SslInfo;

/*
 * Custom blocks.
 */
#define SslInfo_val(v)   ((SslInfo *) Data_custom_val(v))

static int ssl_info_compare(value v1, value v2)
{
    SslInfo *info1 = SslInfo_val(v1);
    SslInfo *info2 = SslInfo_val(v2);

    return info1->fd == info2->fd ? 0 : info1->fd < info2->fd ? -1 : 1;
}

static long ssl_info_hash(value v)
{
    return (long) SslInfo_val(v);
}

static void ssl_finalize(value v_info)
{
    SslInfo *info;

    info = SslInfo_val(v_info);
    if(info->fd >= 0) {
        closesocket(info->fd);
        info->fd = -1;
    }
}

/*
 * Pass info in a custom block.
 */
static struct custom_operations win_handle_ops = {
    "ssl_socket",
    ssl_finalize,
    ssl_info_compare,
    ssl_info_hash,
    custom_serialize_default,
    custom_deserialize_default
};

static value ssl_info_new(SOCKET fd)
{
    value v = alloc_custom(&win_handle_ops, sizeof(SslInfo), 0, 1);
    SslInfo *info = SslInfo_val(v);
    info->fd = fd;
    return v;
}

/*
 * Say whether SSL is enabled.
 */
value lm_ssl_enabled(value x)
{
    return Val_false;
}

/*
 * Start the SSL server.
 */
value lm_ssl_init(value x)
{
    return Val_unit;
}

/*
 * Create a new socket.
 */
value lm_ssl_socket(value v_keyfile)
{
    SOCKET s;

    /* Open the socket */
    s = socket(PF_INET, SOCK_STREAM, 0);
    if(s < 0)
        failwith("lm_ssl_socket: socket call failed");

    return ssl_info_new(s);
}

/*
 * Reuse an existing socket.
 */
value lm_ssl_serve(value v_fd, value v_keyfile, value v_dhfile)
{
    SOCKET s = Int_val(v_fd);
    return ssl_info_new(s);
}

/*
 * Get an integer describing the socket.
 */
value lm_ssl_fd(value v_info)
{
    SslInfo *info = SslInfo_val(v_info);
    return Val_int(info->fd);
}

/*
 * Bind the socket to an address.
 */
value lm_ssl_bind(value v_info, value v_addr, value v_port)
{
    SslInfo *info;
    struct sockaddr_in sin;

    /* Get the address */
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = *(uint32 *)v_addr;
    sin.sin_port = htons((short) Int_val(v_port));

    /* Perform the bind */
    info = SslInfo_val(v_info);
    if(bind(info->fd, (struct sockaddr *) &sin, sizeof(sin)) < 0) {
        uerr("lm_ssl_bind");
    }

    return Val_unit;
}

/*
 * Get the address of the socket.
 */
value lm_ssl_get_addr(value v_info)
{
    SslInfo *info;
    struct sockaddr_in sin;
    socklen_t size;
    value a, addr;

    /* Get the address */
    info = SslInfo_val(v_info);
    size = sizeof(sin);
    if(getsockname(info->fd, (struct sockaddr *) &sin, &size) < 0) {
        uerr("lm_ssl_get_addr: getsockname");
    }

    /* Allocate the address */
    a = alloc_inet_addr(sin.sin_addr.s_addr);
    Begin_root(a);
    addr = alloc_small(2, 0);
    Field(addr, 0) = a;
    Field(addr, 1) = Val_int(ntohs(sin.sin_port));
    End_roots();
    return addr;
}

/*
 * Listen.
 */
value lm_ssl_listen(value v_info, value v_dhfile, value v_count)
{
    SslInfo *info;

    info = SslInfo_val(v_info);
    listen(info->fd, Int_val(v_count));
    return Val_unit;
}

/*
 * Accept a connection on the socket.
 */
value lm_ssl_accept(value v_info)
{
    SslInfo *info;
    SOCKET fd;
    int code;

    /* Listen on the server socket */
    info = SslInfo_val(v_info);
    enter_blocking_section();
    fd = accept(info->fd, (struct sockaddr *) 0, 0);
    fflush(stderr);
    leave_blocking_section();
    if(fd < 0) {
        uerr("lm_ssl_accept");
    }

    /* Allocate a new struct */
    return ssl_info_new(fd);
}

/*
 * Make a connection on the socket.
 */
value lm_ssl_connect(value v_info, value v_addr, value v_port)
{
    SslInfo *info;
    struct sockaddr_in sin;
    SOCKET fd;
    int code;

    /* Get the address */
    sin.sin_family = AF_INET;
    sin.sin_addr.s_addr = *(uint32 *)v_addr;
    sin.sin_port = htons((short) Int_val(v_port));

    /* Make the connection */
    info = SslInfo_val(v_info);
    enter_blocking_section();
    code = connect(info->fd, (struct sockaddr *) &sin, sizeof(sin));
    leave_blocking_section();
    if(code < 0) {
        uerr("lm_ssl_connect");
    }

    return Val_unit;
}

/*
 * Read some data from the connection.
 */
value lm_ssl_read(value v_info, value v_string, value v_off, value v_len)
{
    int off, len, amount;
    SslInfo *info;
    char *buf;

    info = SslInfo_val(v_info);
    buf = String_val(v_string);
    off = Int_val(v_off);
    len = Int_val(v_len);
    enter_blocking_section();
    amount = recv(info->fd, buf + off, len, 0);
    leave_blocking_section();
    return Val_int(amount);
}

/*
 * Write some data to the connection.
 */
value lm_ssl_write(value v_info, value v_string, value v_off, value v_len)
{
    int off, len, amount;
    SslInfo *info;
    char *buf;

    info = SslInfo_val(v_info);
    buf = String_val(v_string);
    off = Int_val(v_off);
    len = Int_val(v_len);
    enter_blocking_section();
    amount = send(info->fd, buf + off, len, 0);
    leave_blocking_section();
    return Val_int(amount);
}

value lm_ssl_flush(value v_info)
{
    return Val_unit;
}

/*
 * Shutdown the connection.
 */
value lm_ssl_shutdown(value v_info)
{
    return Val_unit;
}

/*
 * Close the connection.
 */
value lm_ssl_close(value v_info)
{
    ssl_finalize(v_info);
    return Val_unit;
}

#endif /* !SSL_ENABLED */

File: src/libmojave-external/util/lm_ssl.ml
(*
 * Interface to Open SSL.
 *
 * ----------------------------------------------------------------
 *
 * @begin[license]
 * Copyright (C) 2004 Mojave Group, Caltech
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation,
 * version 2.1 of the License.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Additional permission is given to link this library with the
 * OpenSSL project's "OpenSSL" library, and with the OCaml runtime,
 * and you may distribute the linked executables.  See the file
 * LICENSE.libmojave for more details.
 *
 * Author: Jason Hickey
 * @email{jyh@cs.caltech.edu}
 * @end[license]
 *)
let eprintf = Printf.eprintf
let eflush out =
   output_char out '\n';
   flush out

open Lm_make_printf

(*
 * Define hooks to C code.
 *)
type t
type ssl = t

exception SSLSigPipe

external socket         : string -> t                        = "lm_ssl_socket"
external bind           : t -> Unix.inet_addr -> int -> unit = "lm_ssl_bind"
external getsockname    : t -> Unix.inet_addr * int          = "lm_ssl_get_addr"
external listen         : t -> string -> int -> unit         = "lm_ssl_listen"
external accept         : t -> t                             = "lm_ssl_accept"
external connect        : t -> Unix.inet_addr -> int -> unit = "lm_ssl_connect"
external shutdown       : t -> unit                          = "lm_ssl_shutdown"
external close          : t -> unit                          = "lm_ssl_close"

(*
 * For restarting from an existing socket.
 *)
external fd             : t -> int                           = "lm_ssl_fd"
external serve          : int -> string -> string -> t       = "lm_ssl_serve"

(*
 * Private functions.
 *)
external lm_ssl_enabled  : unit -> bool                       = "lm_ssl_enabled"
external lm_ssl_init     : unit -> unit                       = "lm_ssl_init"
external lm_ssl_read     : t -> string -> int -> int -> int   = "lm_ssl_read"
external lm_ssl_write    : t -> string -> int -> int -> int   = "lm_ssl_write"
external lm_ssl_flush    : t -> unit                          = "lm_ssl_flush"

(*
 * Initialize.
 *)
let () = lm_ssl_init ()
let enabled = lm_ssl_enabled ()

(*
 * Buffered output channel.
 *)
module OutChannel =
struct
   (* Some buffer type *)
   type t =
      { mutable windex : int;
        buffer : string;
        ssl    : ssl
      }

   type result = unit

   (* Buffer length is 1k *)
   let buf_length = 1 lsl 10

   (* Create the output channel *)
   let create ssl =
      { windex = 0;
        buffer = String.create buf_length;
        ssl = ssl
      }

   (* Flush the output *)
   let flush out =
      let { windex = windex;
            buffer = buf;
            ssl = ssl
          } = out
      in
      let rec flush off =
         if off < windex then
            let amount = lm_ssl_write ssl buf off (windex - off) in
               if amount <= 0 then
                  raise SSLSigPipe;
               flush (off + amount)
      in
         flush 0;
         out.windex <- 0

   (* Close the output channel *)
   let close out =
      flush out;
      close out.ssl

   (*
    * Add a single character.
    *)
   let print_char out c =
      let { windex = windex;
            buffer = buf
          } = out
      in
      let windex' = succ windex in
         buf.[windex] <- c;
         out.windex <- windex';
         if windex' = buf_length then
            flush out

   (*
    * Add a string.
    *)
   let rec output out s off len =
      if len <> 0 then
         let { windex = windex;
               buffer = buf
             } = out
         in
         let amount = min len (buf_length - windex) in
         let windex' = windex + amount in
            String.blit s off buf windex amount;
            out.windex <- windex';
            if windex' = buf_length then
               flush out;
            output out s (off + amount) (len - amount)

   let print_string out s =
      output out s 0 (String.length s)

   (*
    * Formatting functions are ignored.
    *)
   let open_box _ _ = ()
   let open_hbox _ = ()
   let open_vbox _ _ = ()
   let open_hvbox _ _ = ()
   let open_hovbox _ _ = ()
   let close_box _ = ()

   (*
    * These formatting actions are partially handled.
    *)
   let print_cut _ =
      ()

   let print_space out =
      print_char out ' '

   let force_newline out =
      print_char out '\n'

   let print_break _ _ _ =
      ()

   let print_flush out =
      flush out;
      lm_ssl_flush out.ssl

   let print_newline =
      force_newline

   let exit _out =
      ()
end

module Printf = MakePrintf (OutChannel)

type ssl_out = OutChannel.t

let out_channel_of_ssl = OutChannel.create
let output_char = OutChannel.print_char
let output_string = OutChannel.print_string
let output_buffer out buf =
   output_string out (Buffer.contents buf)
let output = OutChannel.output
let flush = OutChannel.flush
let close_out = OutChannel.close

let fprintf = Printf.fprintf

(*
 * Input channel.
 *)
module InChannel =
struct
   type t =
      { mutable rindex : int;
        mutable length : int;
        buffer : string;
        ssl : ssl
      }

   (*
    * Input buffer is 1k.
    *)
   let buf_length = 1 lsl 10

   (*
    * Create the buffer.
    *)
   let create ssl =
      { rindex = 0;
        length = 0;
        buffer = String.create buf_length;
        ssl = ssl
      }

   let close inx =
      close inx.ssl

   (*
    * Fill input.
    *)
   let rec fill inx =
      let { rindex = rindex;
            length = length;
            buffer = buf;
            ssl = ssl
          } = inx
      in
         if length <> 0 && rindex = length then
            begin
               inx.rindex <- 0;
               inx.length <- 0;
               fill inx
            end
         else if length <> buf_length then
            begin
               let amount = lm_ssl_read ssl buf length (buf_length - length) in
                  if amount <= 0 then
                     raise End_of_file;
                  inx.length <- length + amount
            end

   (*
    * Get a character.
    *)
   let rec input_char inx =
      let { rindex = rindex;
            length = length;
            buffer = buf
          } = inx
      in
         if rindex = length then
            begin
               fill inx;
               input_char inx
            end
         else
            let c = buf.[rindex] in
               inx.rindex <- succ rindex;
               c

   (*
    * Get a line.
    *)
   let rec input_line_aux out inx =
      let { rindex = rindex;
            length = length;
            buffer = buf
          } = inx
      in
      let rec collect rindex =
         if rindex = length then
            let filled =
               inx.rindex <- rindex;
               try fill inx; true with
                  End_of_file ->
                     false
            in
               if filled then
                  input_line_aux out inx
               else
                  Buffer.contents out
         else
            let c = buf.[rindex] in
               if c = '\n' then
                  begin
                     inx.rindex <- succ rindex;
                     Buffer.contents out
                  end
               else
                  begin
                     Buffer.add_char out c;
                     collect (succ rindex)
                  end
      in
         collect rindex

   let input_line inx =
      if inx.rindex = inx.length then
         fill inx;
      input_line_aux (Buffer.create 64) inx

   (*
    * Get some input.
    * Try to optimize this case a little, and
    * copy directly into the buffer if possible.
    *)
   let rec really_input_start inx s off len =
      let { rindex = rindex;
            length = length;
            buffer = buf
          } = inx
      in
      let amount = min len (length - rindex) in
      let () =
         String.blit buf rindex s off amount;
         inx.rindex <- rindex + amount
      in
         let off = off + amount in
         let len = len - amount in
            if len <> 0 then
               really_input_middle inx s off len

   and really_input_middle inx s off len =
      if len < buf_length then
         really_input_end inx s off len
      else
         let amount = lm_ssl_read inx.ssl s off len in
         let () =
            if amount <= 0 then
               raise End_of_file
         in
         let off = off + amount in
         let len = len - amount in
            really_input_middle inx s off len

   and really_input_end inx s off len =
      let { rindex = rindex;
            length = length;
            buffer = buf
          } = inx
      in
         if length - rindex < len then
            begin
               fill inx;
               really_input_end inx s off len
            end
         else
            begin
               String.blit buf rindex s off len;
               inx.rindex <- rindex + len
            end

   let really_input = really_input_start
end

type ssl_in = InChannel.t

let in_channel_of_ssl = InChannel.create
let input_char = InChannel.input_char
let input_line = InChannel.input_line
let really_input = InChannel.really_input
let close_in = InChannel.close

(*!
 * @docoff
 *
 * -*-
 * Local Variables:
 * Caml-master: "compile"
 * End:
 * -*-
 *)
